### PR TITLE
Handle namepaths inside JSDoc type expressions a bit better 

### LINKED ISSUE #31298: Odd quickinfo result for jsdoc return tag using module
Fro https://github.com/microsoft/vscode/issues/73363

**TypeScript Version:**  3.5.0-dev.20190507

**Search Terms:**
- jsdoc
- hover
- quickinfo

**Repo**
For the js:

```js
/**
 * @returns {module:@nodefuel/web~Webserver~wsServer} Websocket server object
 */
function foo() { }
```

Hover over `foo`

**Bug:**

The return `tags` `text` contains part of the type plus the closing `}`

```
[Trace  - 11:56:05 AM] Response received: quickinfo (38). Request took 2 ms. Success: true 
Result: {
    "kind": "function",
    "kindModifiers": "",
    "start": {
        "line": 4,
        "offset": 10
    },
    "end": {
        "line": 4,
        "offset": 13
    },
    "displayString": "function foo(): any",
    "documentation": "",
    "tags": [
        {
            "name": "returns",
            "text": ":"
        },
        {
            "name": "nodefuel",
            "text": "/web~Webserver~wsServer} Websocket server object"
        }
    ]
}
```

To be consistent with other jsdoc results, I believe we would expect a single `@return` tag with the text:

```
"text": "Websocket server object"
```

**Playground Link:** 

**Related Issues:**

### PR DESCRIPTION
**Pull request by orta**
Fixes #31298

Adds a way for the parser to move the scanner through a namepath inside JSDoc so that it can continue to parse the rest of the statement correctly and show up right in an editor.

### CODE REVIEW

File: src/compiler/parser.ts [hunk 1]
```diff
@@ -2429,8 +2429,16 @@ namespace ts {
...
             const dotdotdot = parseOptionalToken(SyntaxKind.DotDotDotToken);
+            const moduleSpecifier = parseOptionalToken(SyntaxKind.ModuleKeyword);
             let type = parseTypeOrTypePredicate();
             scanner.setInJSDocType(false);
+            if (moduleSpecifier) {
+                const moduleTag = createNode(SyntaxKind.JSDocNamepathType, moduleSpecifier.pos) as JSDocNamepathType;
+                while (token() !== SyntaxKind.CloseBraceToken && token() !== SyntaxKind.EndOfFileToken) {
+                    nextTokenJSDoc();
+                }
+                type = finishNode(moduleTag);
```
Comment (sandersn) [inline src/compiler/parser.ts]: This whole thing should actually come first and be a completely separate code path. Here we call `parseTypeOrTypePredicate` (and allow prefix `...`) in the namepath case, which is wrong.

```suggestion
                return finishNode(moduleTag);
```

File: src/compiler/parser.ts [hunk 2]
```diff
@@ -2429,8 +2429,16 @@ namespace ts {
         function parseJSDocType(): TypeNode {
             scanner.setInJSDocType(true);
             const dotdotdot = parseOptionalToken(SyntaxKind.DotDotDotToken);
+            const moduleSpecifier = parseOptionalToken(SyntaxKind.ModuleKeyword);
             let type = parseTypeOrTypePredicate();
             scanner.setInJSDocType(false);
+            if (moduleSpecifier) {
+                const moduleTag = createNode(SyntaxKind.JSDocNamepathType, moduleSpecifier.pos) as JSDocNamepathType;
+                while (token() !== SyntaxKind.CloseBraceToken && token() !== SyntaxKind.EndOfFileToken) {
```
Comment (sandersn) [inline src/compiler/parser.ts]: I realised that parseJSDocType is also called from parseJSDocParameter, not just single type expressions. That means its right terminators include `,` and `)` See the `ParsingContext.JSDocParameters` entry in `isListTerminator`.

That also means that we need tests for jsdoc-style functions:

```js
/** @type {function(module:xxxxxxxxxx, module:xxxxxxxxxx)} */
```

Comment (sandersn) [reply src/compiler/parser.ts]: I'm not sure whether jsdoc.app mentions whether namepaths can be used in function types, but I think we should support it even if it does not.

**Comment by orta** [comment]
Hrm, these changes seem to affect the behavior of generators somehow - I'm gonna wait till Nathan's back and chat with him about it

File: src/compiler/parser.ts [hunk 3]
```diff
@@ -2431,6 +2431,18 @@ namespace ts {
 
         function parseJSDocType(): TypeNode {
             scanner.setInJSDocType(true);
+            const moduleSpecifier = parseOptionalToken(SyntaxKind.ModuleKeyword);
+            if (moduleSpecifier) {
+                const moduleTag = createNode(SyntaxKind.JSDocNamepathType, moduleSpecifier.pos) as JSDocNamepathType;
+                const terminators = [SyntaxKind.CloseBraceToken, SyntaxKind.EndOfFileToken, SyntaxKind.CommaToken, SyntaxKind.CloseParenToken, SyntaxKind.WhitespaceTrivia];
```
Comment (sandersn) [inline src/compiler/parser.ts]: the new array+member approach feels like it creates unnecessary garbage in a path that may happen a lot.

File: src/compiler/parser.ts [hunk 4]
```diff
@@ -2431,6 +2431,18 @@ namespace ts {
 
         function parseJSDocType(): TypeNode {
             scanner.setInJSDocType(true);
+            const moduleSpecifier = parseOptionalToken(SyntaxKind.ModuleKeyword);
+            if (moduleSpecifier) {
+                const moduleTag = createNode(SyntaxKind.JSDocNamepathType, moduleSpecifier.pos) as JSDocNamepathType;
+                const terminators = [SyntaxKind.CloseBraceToken, SyntaxKind.EndOfFileToken, SyntaxKind.CommaToken, SyntaxKind.CloseParenToken, SyntaxKind.WhitespaceTrivia];
+                while (terminators.indexOf(token()) < 0) {
```
Comment (sandersn) [inline src/compiler/parser.ts]:
```suggestion
                terminate: while (true) {
                    switch(token()) {
                        case SyntaxKind.CloseBraceToken: // etc
                            break terminate;
                    }
```

File: src/compiler/parser.ts (hunk 3 above)
Comment (orta) [reply src/compiler/parser.ts]: Makes sense! Good point.

File: src/compiler/parser.ts (hunk 4 above)
Comment (orta) [reply src/compiler/parser.ts]: I love it, this is the only project I've ever had PR suggestions where people recommend `goto`.

**Comment by orta** [comment]
Will fix that up tomorrow 👍

File: src/compiler/parser.ts (hunk 4 above)
Comment (sandersn) [reply src/compiler/parser.ts]: Gotta go fast!

**Pull request merged**